### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] Last option in exhaustive check is erroring

### ISSUE THREAD

**Issue by blixt** [typescript-eslint/typescript-eslint#4682]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
type Choice = "a" | "b"
const choice: Choice = Math.random() < 0.5 ? "a" : "b"
if (choice === "a") {
    console.log("a!")
} else if (choice === "b") { // error: Unnecessary conditional, both sides of the expression are literal values.
    console.log("b!")
} else {
    assertNever(choice) // this will tell us if we add a third option and don't handle it
}
```

**Expected Result**

No error, because we rely on the literal checks to be exhaustive.

**Actual Result**

The rule causes an error that must be ignored.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.14.0` |
| `@typescript-eslint/parser`        | `5.14.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.11.0` |
| `node`                             | `16` |

**Comment by bradzacher** [comment]
`if` checks aren't a great fit for this sort of exhaustive check - instead just use a `switch`:
```ts
switch (choice) {
  case 'a':
    // ...
    break;

  case 'b':
    // ...
    break;

  default:
    assertNever(choice);
}
```

----

This is working as expected.
It's really, really difficult for us to follow the control-flow analysis of a single variable to determine that a set of `if` checks are intentionally exhaustive.

It's pretty much impossible for us to know that it was specifically the `if` that exhausted the types in the union. Eg:
```ts
if (choice === 'a') {

} else if (otherChoice === 'a') {

} else if (condition) {

} else if (choice === 'b') {
// how do we backtrack up the entire AST to know for certain this check is intentionally exhausitve???
}
```
There are also cases like:
```ts
if (choice === 'a') {
  return;
}

if (choice === 'b') {
  // intentionally exhaustive or logic error?
  // how can we tell?
  return;
}
```

**Comment by bradzacher** [comment]
The reason I say that a switch is better is because you can do away with this sort of "assert never" pattern altogether thanks to rules like [`switch-exhaustiveness-check`](https://typescript-eslint.io/rules/switch-exhaustiveness-check) which will enforce exhaustiveness.

---

> For example an option to disable "both literals are the same" errors

It's more often that you don't want unnecessary checks in your codebase than you do.
Adding an option which blanket ignores unnecessary checks against literals is too far-reaching and *will* cause false negatives.

False positives are okay. They're visible. They're quantifiable and suppressable. "Yes I know that this is unnecessary, but I'm doing it on purpose"
False negatives are BAD. They're invisible. They **harm user trust** when they're found. "WTF why didn't the linter tell me this case was unnecessary?"

---

Disable comments are perfectly acceptable to use.
They are designed exactly for this sort of case where you are ***purposely*** violating a lint rule.

```ts
} else if (
  // eslint-disable-next-line @typescript-eslint/no-unnnecessary-condition -- intentionally unnecessary check for exhaustive purposes
  choice === 'b'
) {
  // ...
}
```
